### DISCUSSION TITLE
installation on Mac with Intel fortran

### DISCUSSION THREAD

**Discussion by youngung** [neperfepx/FEPX#33] [Q&A] [answered]
Thanks for the great software and managing it as an open-sourced project. I wanted to test-run FEPX by installing it on my Mac but I failed to build it. For your information, while I have both gfortran and intel fortran compilers installed on my system, it seems that FEPX picks up Intel Fortran.

The first problem for me was due to the location of variable declarations. I came across below messages over the course of build.

> _------------------------------------------------------------------
>  ~/repo/FEPX/src/build | on main ▓▒░ cmake ..                                                                                                                                                ░▒▓ ✔ | latest Py | 2.4.1 Ruby | at 14:01:24 
> -- The Fortran compiler identification is Intel 2021.1.0.20201112
> -- Detecting Fortran compiler ABI info
> -- Detecting Fortran compiler ABI info - done
> -- Check for working Fortran compiler: /opt/intel/oneapi/compiler/2021.1.1/mac/bin/intel64/ifort - skipped
> -- Checking whether /opt/intel/oneapi/compiler/2021.1.1/mac/bin/intel64/ifort supports Fortran 90
> -- Checking whether /opt/intel/oneapi/compiler/2021.1.1/mac/bin/intel64/ifort supports Fortran 90 - yes
> -- Found MPI_Fortran: /usr/local/Cellar/open-mpi/4.1.1_2/lib/libmpi_usempif08.dylib (found version "3.1") 
> -- Found MPI: TRUE (found version "3.1")  
> -- Fortran compiler: /opt/intel/oneapi/compiler/2021.1.1/mac/bin/intel64/ifort, Intel
> -- Configuring Makefiles for the FEPX software package...
> -- Built date: (2021-07-21)
> -- OS: Darwin 20.5.0
> -- Build Type: Release
> -- Version: 
> -- Preparing library: libf95
> -- Preparing library: libparallel
> -- Preparing library: libfepx
> -- Switching to MPI Fortran compiler: /usr/local/bin/mpif90, Intel
> -- Preparing executable: fepx
> -- FEPX is set to be installed in: /usr/local/bin
> -- Configuring done
> -- Generating done
> -- Build files have been written to: /Users/youngung/repo/FEPX/src/build
>  ~/repo/FEPX/src/build | on main ▓▒░ make                                                                                                                                         ░▒▓ ✔ | took 26s | latest Py | 2.4.1 Ruby | at 14:01:53 
> Scanning dependencies of target libf95
> [  2%] Building Fortran object libf95/CMakeFiles/libf95.dir/intrinsic_types_mod.f90.o
> ifort: command line warning #10006: ignoring unknown option '-ffpe-summary=none'
> [  4%] Building Fortran object libf95/CMakeFiles/libf95.dir/files_mod.f90.o
> ifort: command line warning #10006: ignoring unknown option '-ffpe-summary=none'
> [  6%] Building Fortran object libf95/CMakeFiles/libf95.dir/strings_mod.f90.o
> ifort: command line warning #10006: ignoring unknown option '-ffpe-summary=none'
> [  8%] Building Fortran object libf95/CMakeFiles/libf95.dir/command_mod.f90.o
> ifort: command line warning #10006: ignoring unknown option '-ffpe-summary=none'
> [ 10%] Building Fortran object libf95/CMakeFiles/libf95.dir/timer_mod.f90.o
> ifort: command line warning #10006: ignoring unknown option '-ffpe-summary=none'
> [ 12%] Building Fortran object libf95/CMakeFiles/libf95.dir/libf95.f90.o
> ifort: command line warning #10006: ignoring unknown option '-ffpe-summary=none'
> [ 14%] Linking Fortran static library ../lib/libf95dplab.a
> [ 14%] Built target libf95
> Scanning dependencies of target libparallel
> [ 17%] Building Fortran object libparallel/CMakeFiles/libparallel.dir/parallel_mod.f90.o
> ifort: command line warning #10006: ignoring unknown option '-ffpe-summary=none'
> [ 19%] Building Fortran object libparallel/CMakeFiles/libparallel.dir/gather_scatter_mod.f90.o
> ifort: command line warning #10006: ignoring unknown option '-ffpe-summary=none'
> [ 21%] Linking Fortran static library ../lib/libparallel.a
> [ 21%] Built target libparallel
> Scanning dependencies of target libfepx
> [ 23%] Building Fortran object libfepx/CMakeFiles/libfepx.dir/dimensions_mod.f90.o
> ifort: command line warning #10006: ignoring unknown option '-ffpe-summary=none'
> [ 25%] Building Fortran object libfepx/CMakeFiles/libfepx.dir/quadrature_mod.f90.o
> ifort: command line warning #10006: ignoring unknown option '-ffpe-summary=none'
> [ 27%] Building Fortran object libfepx/CMakeFiles/libfepx.dir/convergence_mod.f90.o
> ifort: command line warning #10006: ignoring unknown option '-ffpe-summary=none'
> [ 29%] Building Fortran object libfepx/CMakeFiles/libfepx.dir/shape_2d_mod.f90.o
> ifort: command line warning #10006: ignoring unknown option '-ffpe-summary=none'
> [ 31%] Building Fortran object libfepx/CMakeFiles/libfepx.dir/surface_mod.f90.o
> ifort: command line warning #10006: ignoring unknown option '-ffpe-summary=none'
> [ 34%] Building Fortran object libfepx/CMakeFiles/libfepx.dir/read_input_mod.f90.o
> ifort: command line warning #10006: ignoring unknown option '-ffpe-summary=none'
> [ 36%] Building Fortran object libfepx/CMakeFiles/libfepx.dir/shape_3d_mod.f90.o
> ifort: command line warning #10006: ignoring unknown option '-ffpe-summary=none'
> [ 38%] Building Fortran object libfepx/CMakeFiles/libfepx.dir/matrix_operations_mod.f90.o
> ifort: command line warning #10006: ignoring unknown option '-ffpe-summary=none'
> /Users/youngung/repo/FEPX/src/libfepx/matrix_operations_mod.f90(1294): error #6415: This name cannot be assigned this data type because it conflicts with prior uses of the name.   [DIM]
>     INTEGER, INTENT(IN) :: DIM
> ---------------------------^
> /Users/youngung/repo/FEPX/src/libfepx/matrix_operations_mod.f90(1797): error #6415: This name cannot be assigned this data type because it conflicts with prior uses of the name.   [ESUB]
>     INTEGER :: ESUB
> ---------------^
> /Users/youngung/repo/FEPX/src/libfepx/matrix_operations_mod.f90(1798): error #6415: This name cannot be assigned this data type because it conflicts with prior uses of the name.   [ESUP]
>     INTEGER :: ESUP
> ---------------^
> compilation aborted for /Users/youngung/repo/FEPX/src/libfepx/matrix_operations_mod.f90 (code 1)
> make[2]: *** [libfepx/CMakeFiles/libfepx.dir/matrix_operations_mod.f90.o] Error 1
> make[1]: *** [libfepx/CMakeFiles/libfepx.dir/all] Error 2
> make: *** [all] Error 2
> 
> -------------

Apparently, variable declarations on DIM, ESUB, ESUP seem to have issues with INTEL Fortran. Moving those lines ahead of other variables in the relevant locations (in source files conjugate_gradient_mod.f90 and matrix_operations_mod.f90 )resolved the issue.

Nevertheless, I was not able to complete the build. When I resume build (using $make ) the building of object FEPX.f90.o failed with the following message.

> [ 57%] Linking Fortran static library ../lib/libfepx.a
> [ 95%] Built target libfepx
> Scanning dependencies of target fepx
> [ 97%] Building Fortran object CMakeFiles/fepx.dir/fepx.f90.o
> gfortran: error: unrecognized command-line option '-module'; did you mean '-Mmodules'?
> make[2]: *** [CMakeFiles/fepx.dir/fepx.f90.o] Error 1
> make[1]: *** [CMakeFiles/fepx.dir/all] Error 2
> make: *** [all] Error 2

It seems that gfortran is used (despite that Intel fortran seems to have been used when building other objects), which resulted in the error message 
> gfortran: error: unrecognized command-line option '-module'; did you mean '-Mmodules'?

How can I resolve this issue? What would you recommend me to try?

**Comment by mkasemer** [comment, marked as answer]
Intel Fortran is currently not tested or supported, as stated in the FEPX manual: we currently only test using GFortran. There is no inherent reason why you cannot use ifort, though you will have to determine the correct compiler options to use, and edit the CMake files accordingly. I would be happy to review any changes made if you would like to try.

**Reply by youngung** [reply]
I would like to contribute. Although I am familiar with make, I have no experience with cmake. So I am not sure where I should start looking at first. Where would you recommend me to look at to get the problematic object compile with Intel compiler?

**Reply by mkasemer** [reply]
@youngung 

Thanks for offering: the largest hurdle for testing with ifort is having ifort in the first place.

You will need to edit the file `src/CMakeLists.txt`, which contains definitions for ifort compiler options. They are currently commented out, as you can see. You can see examples for GFortran throughout, which should allow you to identify appropriate compiler options.

**Reply by mkasemer** [reply]
Some possibilities, from a previous attempt on another system (there were other issues why we could not test ifort there, fully):
```
ELSEIF(CMAKE_Fortran_COMPILER_ID MATCHES Intel)
   set(CMAKE_Fortran_MODDIR_FLAG   "-module ")
#   set(CMAKE_Fortran_FLAGS         "")
   set(CMAKE_Fortran_FLAGS_DEBUG   "-O0 -g -traceback -check bounds -warn all")
   set(CMAKE_Fortran_FLAGS_RELEASE "-O3 -unroll")
endif()
```

**Reply by youngung** [reply]
I still could not build it using Intel Fortran. I would like you to look at the error message from "gfortran". I saw Intel Fortran is picked up as the compiler so I am not sure why the message from 'gfortran' command pops up. Maybe, it is related with mpi not being somehow working with Intel Fortran?

I opted to suppress Intel Fortran on my system to enforce the use of gfortran, then the build is completed.

**Reply by mkasemer** [reply]
@youngung: it is possible that the compiler options I suggested above would not be correct, they may have to be iterated upon.

Concerning the use of gfortran, despite specifying ifort compilers: I'm not sure. Perhaps it has something to do with CMake switching to the use of `mpif90` as your compiler:
```
-- Switching to MPI Fortran compiler: /usr/local/bin/mpif90, Intel
```
I know that it says `Intel`, but is `mpif90` built against `GCC`?

This is all beside the point if you are happy to run using a gfortran compiled version. I am not sure what gains - if any - you would receive from compiling with ifort, we have taken no care to optimize the code in that direction.

Please let me know if you have any other questions, we'd be happy to help.

**Reply by heliolei** [reply]
> I still could not build it using Intel Fortran. I would like you to look at the error message from "gfortran". I saw Intel Fortran is picked up as the compiler so I am not sure why the message from 'gfortran' command pops up. Maybe, it is related with mpi not being somehow working with Intel Fortran?
> 
> I opted to suppress Intel Fortran on my system to enforce the use of gfortran, then the build is completed.

This is useful tip. I try to install FEPX on kubuntu, it has the same compiling problem. I also installed intel fortran. 
To solve this, I suppress the Intel fortran in my bashrc config (simplily comment the intel fortran config). and now it complies well.
Thank you for this tip
